rfcs/0002-validation/text.md

```diff
@@ -0,0 +1,149 @@
+   The outcome of this phase is a file in a format the data portal accepts that follows the schema.
+2. The submitter needs to then upload the dataset file to the portal. This involves things like creating a collection
+   and performing whatever consent/attestation is required. Then the submitter actually uploads the file, a process
+   which is described in detail in other RFCs.
+3. The uploaded file then needs to be validated and processed so it can be available via the portal. This involves four
+   main tasks:
+   - Validating that the file follows the schema.
+   - Updating the portal database with information about the uploaded file, things like tissue or ethnicity that come
+     from the schema.
+   - Converting the uploaded file to different formats and associating those with the dataset.
+   - Creating a cellxgene deployment for the dataset.
+
+This RFC is concerned with phase 3.
+
+### Relevant Components
+
+There are six main components involved in validation and processing:
+
+![High-level Validation Architecture](imgs/validation_architecture.png)
+
+1. Staging bucket. The S3 bucket where uploaded file are placed by the upload process.
+2. Artifact bucket. The S3 bucket where artifacts associated with datasets are stored and served.
+3. cellxgene deployment bucket. The S3 bucket configured to serve datasets via cellxgene.
+4. Portal database. The database where information about datasets is stored. The portal API queries this database when
+   responding to requests.
+5. Validation and processing image. The Docker image containing the dependencies and script to validate and process.
+6. Validation and processing Fargate task. The task definition for executing the image with AWS Fargate.
+
+#### Portal Database Schema
+
+The relevant part of the portal database is three tables: `dataset`, `artifacts`, and `deployment_directories`.
+
+![Database Schema](imgs/db_schema.png)
+
+The `dataset` table tracks the status of datasets and various metadata about them that is used by the portal frontend.
+There are three status fields:
+
+- upload\_status: uploading, uploaded, canceled
+- validation\_status: validating, valid, invalid
+- artifact\_status: creating, complete
+
+The `artifacts` table tracks the different downloadable files associated with a dataset. This stores the format of the
+download and the bucket and object key in the artifact bucket.
+
+The `deployment_directories` table tracks the cellxgene deployments associated with a dataset and the URL where the
+deployment can be accessed.
+
+### Handoff from Upload
+
+At the end of the upload process there must be two elements present in AWS:
+
+1. A row in the `dataset` table with `id` and `collection_id` populated and `upload_status` set to "uploaded".
+2. A dataset file in the staging bucket with a `x-amz-meta-dataset-id` metadata field whose value is the `id` of the new
```

Comment:
Does this need to go into the upload RFC?

Reply:
Yeah. I think I'll need that metadata right? Since the event is coming straight from S3, I don't think the validation task will know how to find the right DB row on its own.

Reply:
The plan was to store the file under {submission_id}/{dataset_id}/{file_name} in the upload bucket. That information should be available in the S3 event, but we can also add that metadata tag if it makes things easier.